torchmultimodal/modules/losses/vqvae.py

```diff
@@ -0,0 +1,28 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from dataclasses import dataclass
+from typing import Any
+
+from torch import nn, Tensor
+
+
+@dataclass
+class CommitmentLossOutput:
+    loss: Tensor
+
+
+class CommitmentLoss(nn.Module):
+    def __init__(self, commitment_cost: float = 1.0, **kwargs: Any):
+        super().__init__()
+        self.mse_loss = nn.MSELoss(reduction="mean")
+        self.commitment_cost = commitment_cost
+
+    def forward(self, quantised: Tensor, encoded: Tensor):
+        # Quantised vectors must be detached because commitment loss only lets gradient flow through encoder output
```

Comment:
noob q, how are the embeddings learnt ? is there another loss fn?

Reply:
The loss consists of 3 parts: 
- reconstruction (delta between input and output tensors)
- VQ objective loss (to move the quantized vectors closer to output of the encoder)
- commitment loss (to regularize the quantized vectors)

Eq 3) in the [paper](https://papers.nips.cc/paper/2017/file/7a98af17e63a0ac09ce2e96d03992fbc-Paper.pdf)

Reply:
Yup, the commitment loss trains the encoder and the VQ objective loss (I've also seen it called codebook loss) is also very similar except it only allows gradient flow to the embedding to train the embedding weights. 

According to the paper, both are used instead of just one combined loss because technically there's no gradient for the argmin of distance to embedding, so they use a straight through estimator which blocks gradients for the embedding.

Which reminds me that I forgot to do the "straight through" estimator... will fix that